// @ota-meshi/eslint-plugin-svelte/lib/rules/dollar-prefixed-store-uses-vars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const ast_utils_1 = require("../utils/ast-utils");
exports.default = (0, utils_1.createRule)("dollar-prefixed-store-uses-vars", {
    meta: {
        docs: {
            description: "prevent $-prefixed variables to be marked as unused",
            category: "System",
            recommended: "base",
        },
        schema: [],
        messages: {},
        type: "problem",
    },
    create(context) {
        if (!context.parserServices.isSvelte) {
            return {};
        }
        function processId(node) {
            const variable = (0, ast_utils_1.findVariable)(context, node);
            if (!variable) {
                return;
            }
            for (const reference of variable.references) {
                if (reference.identifier.name.startsWith("$") &&
                    reference.identifier.name.slice(1) === node.name) {
                    context.markVariableAsUsed(node.name);
                    break;
                }
            }
        }
        return {
            "ImportDefaultSpecifier > Identifier": processId,
            "ImportSpecifier > Identifier.local": processId,
            "ImportNamespaceSpecifier > Identifier": processId,
        };
    },
});
